### ISSUE TITLE
recursivity doesn't place subfolders inside parent folders when using "official" structure

### ISSUE THREAD

**Issue by it-1313** [bluebird75/maildir2mbox#7]
Hello,

first thank you for this script! I had a much harder time than expected to import maildir into Thunderbird without a detour through an imap server, and this script is the one I found!

I use the latest version 1.0 from april 21st of 2020, and I have the "official" maildir format (screenshot)
![Screenshot_20221222_152146](https://user-images.githubusercontent.com/85690326/209162362-b006e9a1-9f56-49c8-aa9a-6bfd65c14ae3.jpg)
When using `maildir2mbox.py -r Maildir_old/ mboxX` (using 22.04 and python 3.10.6), I get
```
maildir2mbox_orig.py -r Maildir_old/ mboxX
ERROR 15:32:49 Missing `new` and/or `cur` subdirectories in path Maildir_old, aborting conversion
15:32:49 Maildir_old/.HO restructuring -> mboxX.sbd/HO restructuring
15:32:49 Processing 77 messages in Maildir_old/.HO restructuring
15:32:49 Maildir_old/.HO restructuring.AS -> mboxX.sbd/HO restructuring.AS #<-- the folder should be calld AS.sbd and be in HO restructuring
15:32:49 Processing 16 messages in Maildir_old/.HO restructuring.AS #<-- the emails should be moved inside HO restructuring/ in a file AS
15:32:49 Done with 1 errors. # can be ignored I think
```

I also tested it with the provided test_data and if I e.g. move the `.coincoin` folder to the same level as the `.INBOX.toto` folder (and rename it to `.INBOX.toto.coucou`) it also wrongly gets placed next to `.INBOX.toto` (and not inside `.INBOX.toto`).

I think the problem lies in `maildir_sub_path` (? or of course, I wrongly apply the script) and I tried to debug it in an online python debugger but had troubles getting the correct variable (path info) of maildir_path to be used. I'm happy to investigate further, e.g. if somebody tells me how I can use a variable depending on a local folder inside an online debugger.

**Comment by intractabilis** [comment]
> I think the problem lies in `maildir_sub_path`

Yes. The problem is there. I am not sure what the intended logic was behind the authors' algorithm, but it doesn't make sense. I am curious how it passed their tests. For example, for Evolution the emails are located in `~/.local/share/evolution/mail/local`. If we run
```
maildir2mbox -r ~/.local/share/evolution/mail/local path/output.mbox
```
then instead of enumerating folders with name starting from `.` inside `~/.local/share/evolution/mail/local`, the algorithm is trying to look for those folders in `~/.local/share/evolution/mail` whose name starts from `local.`.

This is not all. The folder names can have the actual `.`, not as a separator of subfolders. The authors entirely ignored this fact.

**Comment by intractabilis** [comment]
Here is my piece of code that builds the folder hierarchy from the maildir dot separated list of folders. It also takes care of the cases when dot is a part of the name instead of a folder names separator:
```
    # Read the official version of maildir directory structure and store it as
    # (name0level0,
    #     [(name0level1,
    #         [(name0level2,
    #             [])
    #         ]),
    #      (name1level1, [])
    #     ])
    # i.e. it's a list of tuples (a, b), where a is the name of the folder, b is
    # a list of subfolders
    folders_tree = []
    for p in Path(maildir_path).iterdir():
        if p.is_dir() and p.name.startswith('.') and (p / 'cur').exists() and (p / 'new').exists():
            # First we split the name with '.', then we change all sequencies of
            # _NN with characters corresponding to the ASCII code of NN. This
            # way maildir encodes the actual . and _ in the folder names.
            folders = [
                re.sub(
                    R'_[0-9A-Fa-f]{2}'
                  , lambda m :
                        codecs.decode(m.string[m.start(0) + 1:m.end(0)], 'hex').decode("ASCII")
                  , f) for f in p.name.split('.') ][1:]
            # Then we descend down the tree using folder names to find the
            # child.
            cur = folders_tree
            for f in folders:
                next_level = [e for e in cur if e[0] == f]
                if 0 == len(next_level):
                    cur.append((f, []))
                _, cur = [e for e in cur if e[0] == f][0]
```
However, this is not enough, additional modifications to the script needed regarding how it treats levels of hierarchy. For example, let's take Evolution:
```
~/.local/share/evolution/mail/local/
    cur/
    new/
    tmp/
    .Sent/
    .Drafts/
    .Outbox/
    .folder1.folder2
    .folder1.folder2.foder3
    .folder4
```
The script assumes that the folders of the first level (`Sent`, `Outbox`, `Drafts`) are subfolders of the Inbox (the one you specified on the command line):
```
Inbox
    Sent
    Drafts
    Outbox
    folder1
        folder2
            folder3
    folder4
```
However, this is incorrect, they should be on the same level as Inbox. Subfolders of Inbox start from two dots. For example:
```
~/.local/share/evolution/mail/local/
    cur/
    new/
    tmp/
    .Sent/
    .Drafts/
    .Outbox/
    .folder1.folder2
    .folder1.folder2.foder3
    .folder4
    ..folder5
```
Should turn into
```
Inbox
    folder5
Sent
Drafts
Outbox
folder1
    folder2
        folder3
folder4
```
I.e. the script should create several .mbox files on the first level of hierarchy, instead of recursively starting to descend right away in the first folder.

**Comment by intractabilis** [comment]
@it-1313 Here is my complete rewrite. The command line syntax is
```
mailmigration.py maildir_root_folder mbox_output_folder --inboxname=MigratedInbox
```
You need `--inboxname` because the content of the inbox is in the root of the maildir, and it doesn't have its own name in the file system. The output mbox file for the inbox of maildir will be named `MigratedInbox` in the example above.

[mailmigration.tar.gz](https://github.com/bluebird75/maildir2mbox/files/10326303/mailmigration.tar.gz)
